### ISSUE TITLE
Add support for mj-include directive

### ISSUE THREAD

**Issue by pennal** [atrujillofalcon/mjml-rest-client#103]
According to the docs, it is possible to include external files ([see doc here](https://mjml.io/documentation/#mj-include)). Currenlty the `TemplateFactory` does not seem to resolve such tags, but should not be too difficult (Take the path of the input file, and resolve the final path based on the containing folder of the main template)

Could you add support for it? Would do it myself but I have no experience with Kotlin!

**Comment by atrujillofalcon** [comment]
Hi  @pennal . Thanks for reporting this issue. I going to try to implement it .

**Comment by atrujillofalcon** [comment]
@pennal this library use Thymeleaf template engine, so maybe is a good workaround use the Thymeleaf [include](https://www.thymeleaf.org/doc/tutorials/3.0/usingthymeleaf.html#template-layout) functionality.

**Comment by pennal** [comment]
Yep, I agree, that is a workaround. But this means that whatever you design on the website must be adapted once again to work with Thymeleaf. Also, anytime there is an include it must be preprocessed because it may have variables that have the same name as the template they are being included into (I guess you see where I am going with this).

I would love to implement it myself, but I have no knowledge of Kotlin so I am limited in proposing an approach. Whever the user provides a template, you perform a simple extraction of the `<mj-include>` tag with a regex (whipped it up in five minutes, so it definitely needs tuning: `\<(?:mj-include)\s*(?:path=\"(.*)\")\s*(?:\/>|>\<\/mj-include\>)`). For any match you find, you recursively do this (include inside include) and once you reach the bottom you render the templates one by one, replacing the content each time. With this approach you need to add a preprocessing step BEFORE you send out the template to be rendered by Thymeleaf, but it is just another step in the chain.

What I have outlined is a VERY naive method but it may work. Again, I am sorry if I cannot provide any other help....for now I guess I'll use the workaround you provided. Thanks for the library btw!

**Comment by atrujillofalcon** [comment]
Hello @pennal, can you give me more details of how are you using the library? Reading the documentation of MJML I think that the mj-include tag doesn't made for API usage. This library is made for be used on JVM context, rendering Thymeleaf template and consuming the MJML API subsequently.

The main problem with your suggestion is that in order to pre-process the template **it is necessary to create a temporary file** so as not to modify the original file every time the service is invoked. Currently, I provide the file to the Thymeleaf engine and it returns the String, so your solution would lead to mess the library :/

**Comment by pennal** [comment]
Sure.

All of my templates share the same header and footer. This is done for consistency among the different emails. What I do at runtime is simply inject the body, and render it. I have a template for every type of email (reset, confirmation, etc.), but I am repeating both the header/footer and the styling across them. So what I would like to do is something similar to this: extract the header, footer, and main elements to external files, and include them in the different templates. This would allow me to simply edit one file, and have all the changes propagated to all the templates.

I understand this can be achieved using Thymeleaf, but it would essentially repeat what could already be implemented by the library itself. If there is really no chance to do this, I'll implement it using Thymeleaf.

About the temporary files, starting from Java 7 there is an [api for that](https://docs.oracle.com/javase/7/docs/api/java/io/File.html#createTempFile(java.lang.String,%20java.lang.String)). This file gets created in the temp directory and cleared by the OS when not in use, effectively leaving no trace. Instead of doing things in memory, you dump the contents to that file and read/write to it

Hope this explanation clears it up.
